File: python/cache/eviction.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional, Protocol


class EvictionPolicy(Protocol):
    def on_get(self, key: str) -> None:
        ...

    def on_set(self, key: str) -> None:
        ...

    def on_delete(self, key: str) -> None:
        ...

    def evict(self, capacity: int, size: int) -> List[str]:
        ...


@dataclass
class _Node:
    key: str
    prev: Optional["_Node"] = None
    next: Optional["_Node"] = None


class LRUEviction(EvictionPolicy):
    def __init__(self) -> None:
        self._nodes: Dict[str, _Node] = {}
        self._head: Optional[_Node] = None
        self._tail: Optional[_Node] = None

    def on_get(self, key: str) -> None:
        node = self._nodes.get(key)
        if not node:
            return
        self._move_to_head(node)

    def on_set(self, key: str) -> None:
        node = self._nodes.get(key)
        if node:
            self._move_to_head(node)
            return
        node = _Node(key=key)
        self._nodes[key] = node
        self._add_to_head(node)

    def on_delete(self, key: str) -> None:
        node = self._nodes.pop(key, None)
        if not node:
            return
        self._remove(node)

    def evict(self, capacity: int, size: int) -> List[str]:
        evicted: List[str] = []
        while size > capacity and self._tail:
            key = self._tail.key
            evicted.append(key)
            self.on_delete(key)
            size -= 1
        return evicted

    def _add_to_head(self, node: _Node) -> None:
        node.prev = None
        node.next = self._head
        if self._head:
            self._head.prev = node
        self._head = node
        if not self._tail:
            self._tail = node

    def _remove(self, node: _Node) -> None:
        if node.prev:
            node.prev.next = node.next
        else:
            self._head = node.next
        if node.next:
            node.next.prev = node.prev
        else:
            self._tail = node.prev
        node.prev = None
        node.next = None

    def _move_to_head(self, node: _Node) -> None:
        if node is self._head:
            return
        self._remove(node)
        self._add_to_head(node)

File: go/cache/eviction.go
package cache

type EvictionPolicy interface {
	OnGet(key string)
	OnSet(key string)
	OnDelete(key string)
	Evict(capacity int, size int) []string
}

type lruNode struct {
	key  string
	prev *lruNode
	next *lruNode
}

type LRUEviction struct {
	nodes map[string]*lruNode
	head  *lruNode
	tail  *lruNode
}

func NewLRUEviction() *LRUEviction {
	return &LRUEviction{nodes: map[string]*lruNode{}}
}

func (l *LRUEviction) OnGet(key string) {
	node, ok := l.nodes[key]
	if !ok {
		return
	}
	l.moveToHead(node)
}

func (l *LRUEviction) OnSet(key string) {
	node, ok := l.nodes[key]
	if ok {
		l.moveToHead(node)
		return
	}
	node = &lruNode{key: key}
	l.nodes[key] = node
	l.addToHead(node)
}

func (l *LRUEviction) OnDelete(key string) {
	node, ok := l.nodes[key]
	if !ok {
		return
	}
	delete(l.nodes, key)
	l.remove(node)
}

func (l *LRUEviction) Evict(capacity int, size int) []string {
	evicted := []string{}
	for size > capacity && l.tail != nil {
		key := l.tail.key
		evicted = append(evicted, key)
		l.OnDelete(key)
		size--
	}
	return evicted
}

func (l *LRUEviction) addToHead(node *lruNode) {
	node.prev = nil
	node.next = l.head
	if l.head != nil {
		l.head.prev = node
	}
	l.head = node
	if l.tail == nil {
		l.tail = node
	}
}

func (l *LRUEviction) remove(node *lruNode) {
	if node.prev != nil {
		node.prev.next = node.next
	} else {
		l.head = node.next
	}
	if node.next != nil {
		node.next.prev = node.prev
	} else {
		l.tail = node.prev
	}
	node.prev = nil
	node.next = nil
}

func (l *LRUEviction) moveToHead(node *lruNode) {
	if node == l.head {
		return
	}
	l.remove(node)
	l.addToHead(node)
}
